qiime/core/type/signature.py

```diff
@@ -157,11 +197,25 @@ def _parse_signature(self, callable, inputs, parameters, outputs):
                                 (len(outputs), len(output_views)))
 
             for (name, qiime_type), view_type in zip(outputs, output_views):
-                annotated_outputs[name] = ParameterSpec(qiime_type=qiime_type,
-                                                        view_type=view_type)
+                description = output_descriptions.pop(name,
+                                                      ParameterSpec.NOVALUE)
+                annotated_outputs[name] = ParameterSpec(
+                    qiime_type=qiime_type, view_type=view_type,
+                    description=description)
         else:
             for name, qiime_type in outputs:
-                annotated_outputs[name] = ParameterSpec(qiime_type=qiime_type)
+                description = output_descriptions.pop(name,
+                                                      ParameterSpec.NOVALUE)
+                annotated_outputs[name] = ParameterSpec(
+                    qiime_type=qiime_type, description=description)
+
+        # we should have popped the descriptions empty by this point
+        if input_descriptions or parameter_descriptions or output_descriptions:
+            raise TypeError(
+                "Callable does not have parameter(s)/output(s) found in "
+                " descriptions: %r" % list({**input_descriptions,
+                                            **parameter_descriptions,
+                                            **output_descriptions}))
```

Comment:
although, `list(itertools.chain(input_descriptions, parameter_descriptions, output_descriptions))` would probably make more sense, since the key-value relationship isn't important to preserve.